test/alternator: add some Alternator-over-HTTPS tests

nyh opened the pull request:
This patch adds a few tests for Alternator over HTTPS (encrypted HTTP, a.k.a. TLS or SSL). The tests are skipped unless run with "--https", so they will not be run in CI. Nevertheless, they are useful to improve a our understanding on how DynamoDB works over HTTPS and can be a basis for adding more tests for HTTPS support. The included tests pass on both Alternator and AWS DynamoDB.

One test checks that both TLS 1.2 and TLS 1.3 are properly supported, and if chosen by the client, are actually honored. The same test also checks that TLS 1.1 is not supported, and results with a proper error if attempted. Both AWS DynamoDB and Alterator support the same protocols.

Another test verifies that HTTP (unencrypted) requests cannot be sent over an HTTPS port. This is important for security - an installation that chooses to allow only HTTPS wants users to only use encrypted connections, and would not want users to continue sending unencrypted requests to the HTTPS port.

scylladb-promoter commented:
#### :green_circle: CI State: SUCCESS

:white_check_mark: - [Framework test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/16485/)
:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/16485/)
:white_check_mark: - [Unit Tests Custom](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/16485/)
**The following new/updated tests ran 100 times for each mode:**
:small_blue_diamond: alternator
#### Build Details:
- Duration: 2 hr 43 min
- Builder: spider9.cloudius-systems.com

nyh commented:
ping for review and/or merge

ScyllaPiotr commented:
Sorry for the delay, will do this today.

nyh commented on test/alternator/test_https.py:
```diff
@@ -0,0 +1,103 @@
+        # Check that the negotiated TLS version, res.connection.sock.version(),
+        # is the one we tried to set. Note that version() returns a string
+        # like TLSv1.2, while the ssl.TLSVersion attribute name we used for
+        # tls_version uses a underscore.
```
I wish I could, but ssl.TLSVersion - part of the library I can't change - uses names like `TLSv1_1`.

ScyllaPiotr replied:
OK, now I noticed that 2 separate places, both foreign to your code, use different format.

nyh commented on test/alternator/test_https.py:
```diff
@@ -0,0 +1,103 @@
+            # still inspect this connection and its negoatiated TLS version.
+            res = pool.request('GET', https_url, preload_content=False)
+        except urllib3.exceptions.MaxRetryError as e:
+            if support_required or 'NO_PROTOCOLS_AVAILABLE' not in str(e):
```
what do you mean exotic combinations? there is just one boolean operator here, 'or'. In all languages or has low precedence over other non-boolean operators, so in C you do something like `if (a<7 || b<8)` and not `if ((a<7) || (b<8)`. It's not different here.

For the not-required protocols, namely 1.1, it means that yes - it is fine if Scylla doesn't support them, but if it doesn't support them, it must report this error in a clean way (`NO_PROTOCOLS_AVAILABLE`). This is not a theoretical problem - we actually had such a bug, in CQL - this is issue #8827. In that bug, TLS 1.1 was rejected in a bad way. The the CQL version of this test, in test/cqlpy/test_ssl.cc, says it reproduces that issue. Here I didn't want to write I "reproduce" that issue because that original issue was about CQL. But I guess it is "inspired" by that issue.

nyh replied:
I'll update the comment in front of this test to hopefully explain this better (and refer to #8827 as a justification to all of this).

nyh commented on test/alternator/test_https.py:
```diff
@@ -0,0 +1,103 @@
+#############################################################################
+# Tests for Alternator (DynamoDB API) requests over HTTPS (a.k.a. SSL, TLS).
+# These tests are skipped when the tests are *not* running over https, so
+# run the tests with the "--https" parameter to enable them.
```
The "--https" flag of test/alternator/run does a bunch of things - it changes the Scylla configuration of which port to listen on and sets up certificates files. The other thing it does is indeed to use an `https://` URL.

Indeed the test itself recognizes if it connects via https or http according to the URL - if it begins with https:// or not.

So I'm not sure what you found not clear, but I didn't want to explain how test/alternator/run works in this file. Also note that using test/alternator/run is optional (although very easy) - if you really want, you can run your own Scylla installation, and then give its URL to this test, and the test will see if it uses `https://` and only if it is, it can run the tests in this file.

ScyllaPiotr replied:
I suggest to add the supplementary line to the comment:
_The `--https` flag, among other things, prepends the address with `https://`_
so condition in line 19 will becomes coherent with an error in line 20.

nyh replied:
Done (with slightly different words).

nyh commented on test/alternator/test_https.py:
```diff
@@ -0,0 +1,103 @@
+def test_http_on_https(https_url):
+    without_schema = https_url.removeprefix('https://')
+    # If the https url had no port number, it defaults to 443.
+    if ':' not in without_schema:
```
No, it can't. The ':' has special significance in the URL syntax, there is one after the schema (e.g., https) which we removed above, and one before the port (if there is a port). There can't be a colon inside the host name. RFC 1034 (https://datatracker.ietf.org/doc/html/rfc1034) says domain names must contain letters, digits, or hyphen, that's it.

nyh replied:
> Have you checked the [URL RFC](https://datatracker.ietf.org/doc/html/rfc1738#section-3.1)? 

It says
> `//<user>:<password>@<host>:<port>/<url-path>`

and that host is:
> a sequence of domain labels separated by ".", each domain label starting and ending with an alphanumerical character and possibly also containing "-" characters.

So no colons.

> Also, the `url-path` is not forbidden to use `:` if I'm correct.

I think it is forbidden and needs to be URL-encoded, but I am not 100% sure here (I don't remember the entire RFC by heart, only parts of it ;-)), but in any case, as I noted elsewhere, there is no path part here. This isn't some generic URL-parsing function, it's only for the specific URLs used for Alternator and DynamoDB, and these don't have path parts (I agree it's not beautiful how I just assume this).
> 
> My suggestion is not to use the url in the name of the variable, since it's just a protocol, server name/address and port, that's all.

ScyllaPiotr replied:
To put in other words, `https_url` is part of the function interface. Let's prevent some future user (by using better naming or a comment) from providing any legal URL to it only to learn that user name, password and path are not supported.

nyh replied:
`https_url` is nothing more than boto3's `dynamodb.meta.client._endpoint.host`, in practice it's always a schema and a hostname, and nothing more. The boto source code https://github.com/boto/botocore/blob/develop/botocore/endpoint.py calls it "The fully qualified endpoint hostname." but in practice, it can take a schema and also a port number, which is why I called it a "url".

I agree that if I call it a "url", it will be better to allow it to be any URL, including a path and who knows what, and mess just with the schema. I'll change the code to do that (and use some Python URL parsing library instead of hacking it myself), since it looks important to you :-)

ScyllaPiotr replied:
I believe it is not right to harden the code against impractical cases. That's why I suggested a comment or renaming, so no one actually tries out those impractical cases.

nyh replied:
By the way, the Python urllib.parse.urlparse() is unbelievably stupid, I want to gauge my eyes out :-( It can understand the port number, but doesn't allow you to change it and build a new URL. You can change just the "netloc", which contains username, password, hostname and port, together.
I'm still stuggling to make this 100% general.

nyh replied:
If you're curious, the final conversion code:
```python
    p = urllib.parse.urlparse(https_url)
    p = p._replace(scheme='http')
    if p.port is None:
        p = p._replace(netloc=p.netloc+':443')
    http_url = p.geturl()
```
We were lucky that the port is the very last part of the netloc, so if it's missing we can just add `:443` at the end, and no need to do secondary parsing of the netloc - which may include a username, password and hostname.

nyh replied:
> I believe it is not right to harden the code against impractical cases. That's why I suggested a comment or renaming, so no one actually tries out those impractical cases.

It has a hostname, a port number and a scheme. If it walks like a URL and quacks like a URL, it's a URL. I don't have a better name for it. The name boto gives it, "host" is just wrong (because it can also have a scheme and a port).

ScyllaPiotr replied:
Out of curiosity, would this work?
```python 
    (...)
    if p.port is None:
        p.netloc = ""
        p.port = 443
    http_url = p.geturl()
```

The excersice is to check whether `p.hostname` would be used when `p.netloc` is empty string.

[Source](https://docs.python.org/3/library/urllib.parse.html#url-parsing).

nyh commented on test/alternator/test_https.py:
```diff
@@ -0,0 +1,103 @@
+    without_schema = https_url.removeprefix('https://')
+    # If the https url had no port number, it defaults to 443.
+    if ':' not in without_schema:
+        without_schema = without_schema + ':443'
```
It's not imprecise, it's indeed without schema - the "schema" in URL jargon is the "https://" thing.

I agree that this code assumes there is no path part in the URL. But that's the reality, in both Alternator and DynamoDB. I agree that I could have written this code to work in a more general way that will work in that case too, perhaps using a URL parsing library instead of my own simple string functions. But I just didn't think it was worth the effort - because in practice, this URL will never have a path component.

ScyllaPiotr replied:
My suggestion is not to use the url in the name of the variable, since it's just a protocol, server name/address and port, that's all.

nyh replied:
I see. DynamoDB indeed calls it `dynamodb.meta.client._endpoint.host`, i.e, "host", not "url", but I found this naming even more confusing because it's not just the host, it also has a schema and a port - i.e., it's a bona-fide URL, just without a path part.

nyh replied:
Changed (see the other thread of discussion)

nyh commented on test/alternator/test_https.py:
```diff
@@ -0,0 +1,103 @@
+    # For example, http://dynamodb.us-east-1.amazonaws.com:443    
+    http_url = 'http://' + without_schema
+    with urllib3.PoolManager(retries=0, cert_reqs=ssl.CERT_NONE) as pool:
+        # Sanity check: the original https:// URL works:
```
But anyway, the point of this test is to check that if you have an HTTPS URL, an HTTP request to it will **not** work. A silly way in which a broken implementation can pass this test is that if it fails **every** request, no matter if it's http or https :-) So as a sanity check, we verify that https **is** supported, and only http is rejected.

As you saw, I did wrote a comment about this "sanity check", but maybe you found it too short?

ScyllaPiotr replied:
> A silly way in which a broken implementation can pass this test is that if it fails **every** request, no matter if it's http or https :-) So as a sanity check, we verify that https **is** supported, and only http is rejected.

Agreed. Elevating the sanity check up would be more intuitive. Or changing the function name to _test_https_and_http_on_https_port_. Yeah it's long, but it's more communicative.

ScyllaPiotr replied:
What is the message printed when assert in line 89 fails?

nyh replied:
Just like every assertion failure in pytest, it tells you exactly which line of the code had the assertion which failed, and if the assert was `res.status == 200` it will tell you what res.status is.

I don't expect any developer to just read the assertion failure (if it happens) and immediate understand what happens. The developer is expected to go to the line of the test code that he's told about, and look at what is being checked, and comments.

I'll improve the terse comment about "sanity check" to make it less terse.

nyh commented on test/alternator/test_https.py:
```diff
@@ -0,0 +1,103 @@
+    http_url = 'http://' + without_schema
+    with urllib3.PoolManager(retries=0, cert_reqs=ssl.CERT_NONE) as pool:
+        # Sanity check: the original https:// URL works:
+        res = pool.request('GET', https_url)
```
This `cert_reqs=ssl.CERT_NONE` is not an option the server knows about. It just modifies the **client's** certificate checking: By default, when the server presents a certificate to the client, the client verifies that this certificate was signed by a familiar _certificate authority_. The `CERT_NONE` option tells the client not to bother with testing this. Obviously, a production **client** would never do this, but in the test we need to do this because we run Scylla with "fake" (self-signed) certificates, we don't want to pay a certificate authority to sign certificates for our tests, nor to set up real domain names for these certificates.

ScyllaPiotr replied:
Now it's clear, thanks.
IMO it merits a comment.

ScyllaPiotr replied:
I propose:
`ssl.CERT_NONE` flag causes the client not to verify the server certificate.

nyh replied:
Done (slightly different wording)

nyh commented:
Pushed new version, rebased and changed various things following @ScyllaPiotr's review. In particular, don't try hacky string manipulation on the endpoint URL, and instead use Python's `urllib.parse.urlparse()`

scylladb-promoter commented:
#### :green_circle: CI State: SUCCESS

:white_check_mark: - [Framework test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/16862/)
:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/16862/)
:white_check_mark: - [Unit Tests Custom](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/16862/)
**The following new/updated tests ran 100 times for each mode:**
:small_blue_diamond: alternator
#### Build Details:
- Duration: 1 hr 53 min
- Builder: spider7.cloudius-systems.com

ScyllaPiotr commented on test/alternator/test_https.py:
```diff
@@ -0,0 +1,114 @@
+# security reason for doing this, and doesn't want us to allow an unencrypted
+# HTTP request to be sent over the HTTPS port. So let's verify that an HTTP
+# request on an HTTPS port indeed does not work.
+def test_http_on_https(https_url):
```
It's a bit ambiguous, maybe
`def test_http_on_https_port`
or
`def test_http_on_https_listener`
?

ScyllaPiotr reviewed (approved):
I commented like 2 tiny things, but none of these should actually block the progress of this PR.

nyh commented:
> I commented like 2 tiny things, but none of these should actually block the progress of this PR.

I only found one comment, about the test name "test_http_on_https". I do think the name "test_http_on_https" is **suggestive** enough, with a 9 line comment explaining exactly what this test does. Please remember that unlike usual functions in code, which need to be **called** and should be readable in those call locations, test functions are never called. The test's name doesn't need to reasonable, but not more than that - the comment is more important, and more useful for readers. I hate it when I see somebody went to all the trouble to invent a 70-character test name but I just don't understand what he meant, and there is no comment to explain what he meant.

What is the other comment left open? I can't find it.

ScyllaPiotr commented:
> What is the other comment left open? I can't find it.

There is no other, this was the only one.

nyh commented:
@scylladb/scylla-maint please consider merging.
